File: Data Structures & Algorithms/search-for-word-ii/submission-4.py
class TrieNode:
    __slots__ = ("children", "is_end", "word")

    def __init__(self):
        self.children = {}
        self.is_end = False
        self.word = ""

class Solution:
    def __init__(self) -> None:
        self.root = TrieNode()

    def trie_insert(self, word: str) -> None:
        current = self.root
        for char in word:
            current = current.children.setdefault(char, TrieNode())
        current.is_end = True
        current.word = word

    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        # Metadata
        width, height = len(board[0]), len(board)

        # Build Trie first
        for word in words:
            self.trie_insert(word)

        # DFS: Check if Word exists
        result = set()
        visited = set()
        def dfs(node: TrieNode, row: int, col: int) -> None:
            if node.is_end:
                result.add(node.word)

            directions = ((row + 1, col), (row - 1, col), (row, col + 1), (row, col - 1))
            for r, c in directions:
                if (r < 0 or c < 0 or
                    r == height or c == width or
                    (r, c) in visited or
                    board[r][c] not in node.children
                ):
                    continue

                visited.add((r, c))
                dfs(node.children[board[r][c]], r, c)
                visited.remove((r, c))

        # Go through all Starting-Points
        for row in range(height):
            for col in range(width):
                if board[row][col] in self.root.children:
                    visited.add((row, col))
                    dfs(self.root.children[board[row][col]], row, col)
                    visited.remove((row, col))

        return list(result)

File: Data Structures & Algorithms/valid-sudoku/submission-4.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<int> rows(9);
        vector<int> cols(9);
        vector<int> blocks(9);

        for (int r = 0; r < 9; ++r) {
            for (int c = 0; c < 9; ++c) {
                if (board[r][c] == '.') {
                    continue;
                }

                int b = (r / 3) * 3 + (c / 3);
                int num = board[r][c] - '0';

                // Row Check
                if (rows[r] & (1 << num)) {
                    return false;
                }
                rows[r] |= (1 << num);

                // Col Check
                if (cols[c] & (1 << num)) {
                    return false;
                }
                cols[c] |= (1 << num);

                // Block Check
                if (blocks[b] & (1 << num)) {
                    return false;
                }
                blocks[b] |= (1 << num);
            }
        }

        return true;
    }
};
